pkg/runtime/setup.go

```diff
@@ -287,14 +292,24 @@ func (s *setup) obtain(artifact *buildplan.Artifact) (rerr error) {
 		}
 	} else {
 		// Read the artifact from the archive.
+		if err := s.fireEvent(events.ArtifactDownloadStarted{artifact.ArtifactID, 0}); err != nil {
+			return errs.Wrap(err, "Could not handle ArtifactDownloadStarted event")
+		}
+
 		var err error
-		name := artifact.ArtifactID.String() + s.opts.ArtifactExt
-		artifactFile := filepath.Join(s.opts.FromArchiveDir, name)
-		logging.Debug("Reading file '%s' for '%s'", artifactFile, artifact.DisplayName)
+		name := artifact.ArtifactID.String() + s.opts.FromArchive.ArtifactExt
+		artifactFile := filepath.Join(s.opts.FromArchive.Dir, name)
 		b, err = fileutils.ReadFile(artifactFile)
 		if err != nil {
+			if err2 := s.fireEvent(events.ArtifactDownloadFailure{artifact.ArtifactID, err}); err2 != nil {
+				err = errs.Pack(err, errs.Wrap(err2, "Could not handle ArtifactDownloadFailure event"))
+			}
 			return errs.Wrap(err, "read from archive failed")
 		}
+
+		if err := s.fireEvent(events.ArtifactDownloadSuccess{artifact.ArtifactID}); err != nil {
+			return errs.Wrap(errs.Pack(err, err), "Could not handle ArtifactDownloadSuccess event")
+		}
```

Comment:
I just realized we are in fact handling unpack events separately. So I think it makes more sense that we don't send the download events and instead set the toDownload counter to 0 if we are installing from an archive, ie. https://github.com/ActiveState/cli/blob/6ff1b5ad197179213e3d93cf0ff22c516bff4e4b/pkg/runtime/setup.go#L114-L114

Apologies, I didn't have this in mind with my initial review.

Reply:
Okay, well that complicated things. Let me know if my solution is sub-optimal.